<a href="https://colab.research.google.com/github/TN1928/machinescreatemachines/blob/main/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_1_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("/content/crimes_against_women_2001-2014.csv")


##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [3]:
df.columns #Колонки

Index(['Unnamed: 0', 'STATE/UT', 'DISTRICT', 'Year', 'Rape',
       'Kidnapping and Abduction', 'Dowry Deaths',
       'Assault on women with intent to outrage her modesty',
       'Insult to modesty of Women', 'Cruelty by Husband or his Relatives',
       'Importation of Girls'],
      dtype='object')

In [4]:
df[:5]#Первые 5 значений

Unnamed: 0        STATE/UT       DISTRICT  Year  Rape  \
0           0  ANDHRA PRADESH       ADILABAD  2001    50   
1           1  ANDHRA PRADESH      ANANTAPUR  2001    23   
2           2  ANDHRA PRADESH       CHITTOOR  2001    27   
3           3  ANDHRA PRADESH       CUDDAPAH  2001    20   
4           4  ANDHRA PRADESH  EAST GODAVARI  2001    23   

   Kidnapping and Abduction  Dowry Deaths  \
0                        30            16   
1                        30             7   
2                        34            14   
3                        20            17   
4                        26            12   

   Assault on women with intent to outrage her modesty  \
0                                                149     
1                                                118     
2                                                112     
3                                                126     
4                                                109     

   Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                          34                                  175   
1                          24                                  154   
2                          83                                  186   
3                          38                                   57   
4                          58                                  247   

   Importation of Girls  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0

In [ ]:
df.tail(5)#Последние 5 значений

Unnamed: 0     STATE/UT           DISTRICT  Year  Rape  \
10672         832  Lakshadweep        Lakshadweep  2014     1   
10673         833  Lakshadweep  Total District(s)  2014     1   
10674         834   Puducherry           Karaikal  2014     3   
10675         835   Puducherry         Puducherry  2014     7   
10676         836   Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0

In [ ]:
df.describe() #Статистику

Unnamed: 0          Year          Rape  Kidnapping and Abduction  \
count  10677.000000  10677.000000  10677.000000              10677.000000   
mean    3871.570198   2007.697949     57.989885                 69.888358   
std     2816.369918      4.046874    214.230398                311.623450   
min        0.000000   2001.000000      0.000000                  0.000000   
25%     1009.000000   2004.000000      8.000000                  6.000000   
50%     3678.000000   2008.000000     22.000000                 20.000000   
75%     6347.000000   2011.000000     44.000000                 49.000000   
max     9016.000000   2014.000000   5076.000000              10626.000000   

       Dowry Deaths  Assault on women with intent to outrage her modesty  \
count  10677.000000                                       10677.000000     
mean      20.181699                                         113.539196     
std       98.276531                                         458.903951     
min        0.000000                                           0.000000     
25%        1.000000                                          10.000000     
50%        5.000000                                          34.000000     
75%       16.000000                                          85.000000     
max     2469.000000                                       10001.000000     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
count                10677.000000                         10677.000000   
mean                    27.419313                           209.224314   
std                    167.806797                           905.664362   
min                      0.000000                             0.000000   
25%                      0.000000                            11.000000   
50%                      2.000000                            50.000000   
75%                     12.000000                           144.000000   
max                   4970.000000                         23278.000000   

       Importation of Girls  
count          10677.000000  
mean               0.175330  
std                2.228637  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max               83.000000

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета




1.   **Ответ 1.** 3871.570198	2007.697949	57.989885	69.888358	20.181699	113.539196	27.419313	209.224314	0.175330
2.   **Ответ 2.** min	0.000000	2001.000000	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000

max	9016.000000	2014.000000	5076.000000	10626.000000	2469.000000	10001.000000	4970.000000	23278.000000	83.000000



##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [9]:

df.Year = df.Year.astype("int")
df.DISTRICT = df.DISTRICT.astype("category")


#Вывести типы колонок тут
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10677 entries, 0 to 10676
Data columns (total 11 columns):
 #   Column                                               Non-Null Count  Dtype   
---  ------                                               --------------  -----   
 0   Unnamed: 0                                           10677 non-null  int64   
 1   STATE/UT                                             10677 non-null  object  
 2   DISTRICT                                             10677 non-null  category
 3   Year                                                 10677 non-null  int64   
 4   Rape                                                 10677 non-null  int64   
 5   Kidnapping and Abduction                             10677 non-null  int64   
 6   Dowry Deaths                                         10677 non-null  int64   
 7   Assault on women with intent to outrage her modesty  10677 non-null  int64   
 8   Insult to modesty of Women                           106

Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

DISTRICT, Year - категория;
 Rape - численные


##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [7]:
df[df["Dowry Deaths"] < 5]


#Фильтр колонки 1

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
5               5  ANDHRA PRADESH      GUNTAKAL RLY.  2001     0   
19             19  ANDHRA PRADESH  SECUNDERABAD RLY.  2001     0   
21             21  ANDHRA PRADESH         VIJAYAWADA  2001    25   
22             22  ANDHRA PRADESH    VIJAYAWADA RLY.  2001     1   
23             23  ANDHRA PRADESH      VISAKHA RURAL  2001    12   
...           ...             ...                ...   ...   ...   
10672         832     Lakshadweep        Lakshadweep  2014     1   
10673         833     Lakshadweep  Total District(s)  2014     1   
10674         834      Puducherry           Karaikal  2014     3   
10675         835      Puducherry         Puducherry  2014     7   
10676         836      Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
5                             0             0   
19                            0             1   
21                           48             2   
22                            0             0   
23                           12             3   
...                         ...           ...   
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
5                                                      1     
19                                                     0     
21                                                    84     
22                                                     1     
23                                                    67     
...                                                  ...     
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
5                               0                                    0   
19                              1                                    0   
21                            122                                  520   
22                              1                                    0   
23                             48                                   99   
...                           ...                                  ...   
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
5                         0  
19                        0  
21                        0  
22                        0  
23                        0  
...                     ...  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0  

[4999 rows x 11 columns]

In [10]:
df[df["Year"] < 2002]

#Фильтр колонки 2

Unnamed: 0        STATE/UT       DISTRICT  Year  Rape  \
0             0  ANDHRA PRADESH       ADILABAD  2001    50   
1             1  ANDHRA PRADESH      ANANTAPUR  2001    23   
2             2  ANDHRA PRADESH       CHITTOOR  2001    27   
3             3  ANDHRA PRADESH       CUDDAPAH  2001    20   
4             4  ANDHRA PRADESH  EAST GODAVARI  2001    23   
..          ...             ...            ...   ...   ...   
711         711           DELHI           WEST  2001    45   
712         712     LAKSHADWEEP    LAKSHADWEEP  2001     0   
713         713     LAKSHADWEEP          TOTAL  2001     0   
714         714      PUDUCHERRY    PONDICHERRY  2001     9   
715         715      PUDUCHERRY          TOTAL  2001     9   

     Kidnapping and Abduction  Dowry Deaths  \
0                          30            16   
1                          30             7   
2                          34            14   
3                          20            17   
4                          26            12   
..                        ...           ...   
711                        99            18   
712                         0             0   
713                         0             0   
714                         3             1   
715                         3             1   

     Assault on women with intent to outrage her modesty  \
0                                                  149     
1                                                  118     
2                                                  112     
3                                                  126     
4                                                  109     
..                                                 ...     
711                                                 65     
712                                                  0     
713                                                  0     
714                                                 35     
715                                                 35     

     Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                            34                                  175   
1                            24                                  154   
2                            83                                  186   
3                            38                                   57   
4                            58                                  247   
..                          ...                                  ...   
711                           2                                   22   
712                           0                                    0   
713                           0                                    0   
714                          27                                    3   
715                          27                                    3   

     Importation of Girls  
0                       0  
1                       0  
2                       0  
3                       0  
4                       0  
..                    ...  
711                     0  
712                     0  
713                     0  
714                     0  
715                     0  

[716 rows x 11 columns]

In [11]:
df[df["Kidnapping and Abduction"] > 100]

#Фильтр колонки 3

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
28             28  ANDHRA PRADESH              TOTAL  2001   871   
64             64           ASSAM             NAGAON  2001    51   
70             70           ASSAM              TOTAL  2001   817   
115           115           BIHAR              TOTAL  2001   888   
135           135    CHHATTISGARH              TOTAL  2001   959   
...           ...             ...                ...   ...   ...   
10664         824        Delhi UT              South  2014   280   
10665         825        Delhi UT         South-East  2014   266   
10666         826        Delhi UT         South-West  2014   234   
10670         830        Delhi UT               West  2014   251   
10671         831        Delhi UT  Total District(s)  2014  2096   

       Kidnapping and Abduction  Dowry Deaths  \
28                          765           420   
64                          109             1   
70                         1070            59   
115                         518           859   
135                         171            70   
...                         ...           ...   
10664                       369             6   
10665                       402            24   
10666                       397            17   
10670                       546            13   
10671                      4034           153   

       Assault on women with intent to outrage her modesty  \
28                                                  3544     
64                                                   113     
70                                                   850     
115                                                  562     
135                                                 1763     
...                                                  ...     
10664                                                595     
10665                                                469     
10666                                                494     
10670                                                484     
10671                                               4322     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
28                           2271                                 5791   
64                              4                                  132   
70                              4                                 1248   
115                            21                                 1558   
135                           161                                  840   
...                           ...                                  ...   
10664                         665                                  143   
10665                         123                                  312   
10666                          68                                  318   
10670                          63                                  416   
10671                        1361                                 3173   

       Importation of Girls  
28                        7  
64                        0  
70                        0  
115                      83  
135                       0  
...                     ...  
10664                     0  
10665                     0  
10666                     0  
10670                     0  
10671                     0  

[1123 rows x 11 columns]

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [12]:
df.groupby('Year').mean(numeric_only=True)

#Группировка

Unnamed: 0       Rape  Kidnapping and Abduction  Dowry Deaths  \
Year                                                                  
2001       357.5  44.902235                 40.907821     19.136872   
2002      1075.0  45.543811                 40.350487     18.976356   
2003      1798.5  43.535714                 36.527473     17.054945   
2004      2527.0  50.021948                 42.737997     19.275720   
2005      3258.0  50.092769                 42.974079     18.518417   
2006      3994.5  52.291892                 47.064865     20.589189   
2007      4736.0  55.819650                 54.955585     21.784657   
2008      5488.0  56.417871                 60.286465     21.477004   
2009      6252.0  55.794003                 67.121252     21.859192   
2010      7025.0  56.924262                 76.495507     21.543004   
2011      7810.0  61.203540                 89.924147     21.790139   
2012      8611.0  61.462392                 94.357583     20.303329   
2013       411.0  81.912515                126.077764     19.642770   
2014       418.0  87.777778                136.943847     20.203106   

      Assault on women with intent to outrage her modesty  \
Year                                                        
2001                                          95.318436     
2002                                          94.417246     
2003                                          90.491758     
2004                                          94.834019     
2005                                          93.246930     
2006                                          98.964865     
2007                                         104.263795     
2008                                         106.210250     
2009                                         100.941330     
2010                                         104.269576     
2011                                         108.642225     
2012                                         111.839704     
2013                                         171.905225     
2014                                         196.499403     

      Insult to modesty of Women  Cruelty by Husband or his Relatives  \
Year                                                                    
2001                   27.223464                           137.346369   
2002                   28.247566                           136.959666   
2003                   33.859890                           139.293956   
2004                   27.437586                           159.454047   
2005                   27.241473                           159.124147   
2006                   26.935135                           170.616216   
2007                   29.475101                           204.387618   
2008                   32.099869                           213.781866   
2009                   28.706649                           233.496741   
2010                   25.573813                           241.440308   
2011                   21.668774                           250.657396   
2012                   22.621455                           262.705302   
2013                   30.592953                           288.860267   
2014                   23.261649                           293.612903   

      Importation of Girls  
Year                        
2001              0.318436  
2002              0.211405  
2003              0.126374  
2004              0.244170  
2005              0.406548  
2006              0.181081  
2007              0.164199  
2008              0.176084  
2009              0.125163  
2010              0.092426  
2011              0.202276  
2012              0.145499  
2013              0.075334  
2014              0.031063

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [ ]:
#Агрегация